Algorithm.Framework/Portfolio/MeanVarianceOptimizationPortfolioConstructionModel.py

```diff
@@ -0,0 +1,165 @@
+        self.minimum_weight = kwargs['minimum_weight'] if 'minimum_weight' in kwargs else -1
+        self.maximum_weight = kwargs['maximum_weight'] if 'maximum_weight' in kwargs else 1
+        self.target_return = kwargs['target_return'] if 'target_return' is kwargs else 2
+        self.optimization_method = kwargs['optimization_method'] if 'optimization_method' in kwargs else self.minimum_variance
+        self.symbolDataBySymbol = {}
+
+
+    def CreateTargets(self, algorithm, insights):
+        """ 
+        Create portfolio targets from the specified insights
+        Args:
+            algorithm: The algorithm instance
+            insights: The insights to create portoflio targets from
+        Returns: 
+            An enumerable of portoflio targets to be sent to the execution model
+        """
+        symbols = [insight.Symbol for insight in insights]
+        if len(symbols) == 0:
+            return []
+
+        for insight in insights:
+            symbolData = self.symbolDataBySymbol.get(insight.Symbol)
+            symbolData.Add(algorithm.Time, insight.Magnitude)
+
+        # Create a dictionary keyed by the symbols in the insights with an pandas.Series as value to create a data frame
+        returns = { str(symbol) : data.Return for symbol, data in self.symbolDataBySymbol.items() if symbol in symbols }
+        returns = pd.DataFrame(returns)
+
+        # The optimization method processes the data frame 
+        opt, weights = self.optimization_method(returns)
+
+        # Create portfolio targets from the specified insights
+        for insight in insights:
+            weight = weights[str(insight.Symbol)]
+            yield PortfolioTarget.Percent(algorithm, insight.Symbol, weight)
+
+
+    def OnSecuritiesChanged(self, algorithm, changes):
+        '''Event fired each time the we add/remove securities from the data feed
+        Args:
+            algorithm: The algorithm instance that experienced the change in securities
+            changes: The security additions and removals from the algorithm'''
+
+        # clean up data for removed securities
+        removed = [ x.Symbol for x in changes.RemovedSecurities ]
+        if len(removed) > 0:
+            for subscription in algorithm.SubscriptionManager.Subscriptions:
+                symbol = subscription.Symbol
+                if symbol in removed and symbol in self.symbolDataBySymbol:
+                    self.symbolDataBySymbol.pop(symbol)
+
+        # initialize data for added securities
+        for added in changes.AddedSecurities:
+            if added.Symbol not in self.symbolDataBySymbol:
+                symbolData = SymbolData(added, self.period)
+                self.symbolDataBySymbol[added.Symbol] = symbolData
+
+                # seed new indicators using history request
+                history = algorithm.History(added.Symbol, self.period, self.resolution)
+                for row in history.itertuples():
```

Comment:
it may be a little messier, but issuing a single history request for all new ROC indicators will be much more efficient that issuing one a a time synchronously.